pkg/agent/util/net_linux.go

```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
 	cmd := exec.Command("ovs-vsctl", "--if-exists", "del-port", brName, portName)
 	return cmd.Run()
 }
+
+func HostInterfaceExists(ifName string) bool {
+	_, err := netlink.LinkByName(ifName)
+	if err == nil {
+		return true
+	}
+	if _, ok := err.(netlink.LinkNotFoundError); ok {
+		return false
+	}
+	klog.ErrorS(err, "Failed to find host interface", "name", ifName)
+	return false
+}
+
+func GetUplinkConfig(uplinkIfName string) (*config.AdapterNetConfig, error) {
+	iface, err := net.InterfaceByName(uplinkIfName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get interface", "uplinkIfName", uplinkIfName)
+		return nil, err
+	}
+	addrs, err := GetGlobalIPNetsByName(iface)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get address", "iface", iface)
+		return nil, err
+	}
+	gw, routes, err := getRoutesAndDefaultGwOnInterface(iface.Index)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get routes", "iface.Index", iface.Index)
+		return nil, err
+	}
+	return &config.AdapterNetConfig{
+		Name:    uplinkIfName,
+		Index:   iface.Index,
+		MAC:     iface.HardwareAddr,
+		IPs:     addrs,
+		Routes:  routes,
+		Gateway: gw,
+		MTU:     iface.MTU,
+	}, nil
+}
+
+func RenameInterface(from, to string) error {
+	var renameErr error
+	// Add retry here because it might return the error of device is busy for one time configuration.
+	pollErr := wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
+		renameErr = renameHostInterface(from, to)
+		if renameErr != nil {
+			klog.InfoS("Unable to rename host interface name, retrying", "oldName", from, "newName", to, "err", renameErr)
+			return false, nil
+		}
+		return true, nil
+	})
+	if pollErr != nil {
+		return fmt.Errorf("failed to rename host interface name %s to %s", from, to)
+	}
+	return nil
+}
+
+func MoveIFConfigurations(ips []*net.IPNet, routes []interface{}, mtu int, from string, to string) error {
```

Comment:
As an util function, it's hard to understand what the function does from the arguments.
If it's a function moving ip configurations from one interface to another, there should be just two arguments represeting the source and destination.
If it's a function setting the provided ip address, routes, mtu to an interface, it doesn't make sense to pass `from` and have a logic to check the existence of `from`.
This business specific steps make the function cannot really be reused by other consumers and the function name may be confusing. I would suggest move some logic to caller side, like separating cleanup of `from` and configuration of `to`.

Reply:
We put it here because windows will also need to invoke this function. But this function should do nothing for windows case. But it is true that the function is hard to reuse. So I move this function to external_node_controller_linux.go and external_node_controller_windows.go in the latest update. Please help review this part, thanks.